### PR TITLE
Introducing Pwl waveform

### PR DESCRIPTION
**Pull request by marcin-golebiowski**
Comments will be added in the next commit (tomorrow)

### CODE REVIEW

File: SpiceSharp/Components/Waveforms/Pwl.cs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        protected List<Point> CreatePoints(double[] times, double[] voltages, double initialVoltage)
+        {
+            var result = new List<Point>();
+
+            for (var i = 0; i < times.Length; i++)
+            {
+                result.Add(new Point(times[i], voltages[i]));
+            }
+
+            if (times[0] != 0.0)
```
[the remark this answers is not in the document]
Comment (svenboulanger) [reply SpiceSharp/Components/Waveforms/Pwl.cs]: Double check behavior with LTSpice or another simulator.

File: SpiceSharp/Components/Waveforms/Pwl.cs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            if (time.Equals(0.0))
+                Value = Points[0].Y;
+
+            if (simulation.Method is IBreakpoints method)
+            {
+                var breaks = method.Breakpoints;
+                if (!method.Break)
+                    return;
+
+                for (int i = 0; i < Points.Count; i++)
```
Comment (svenboulanger) [inline SpiceSharp/Components/Waveforms/Pwl.cs]: Every time a breakpoint is hit, you add all points again to the breakpoint list. Either add the breakpoints only once or only add the next breakpoint. The second option is preferred if possible.

File: SpiceSharp/Components/Waveforms/Pwl.cs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+            {
+                result.Insert(0, new Point(0.0, initialVoltage));
+            }
+
+            return result;
+        }
+
+        protected static double GetLineValue(List<Point> points, LineDefinition[] lines, double x)
+        {
+            int index = 0;
```
Comment (svenboulanger) [inline SpiceSharp/Components/Waveforms/Pwl.cs]: Can be optimized: When a timepoint is accepted you can assume that there will not be an `x` smaller than than the accepted timepoint. So you can find an index > 0 to start searching from.

File: SpiceSharp/Components/Waveforms/Pwl.cs [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                throw new NullReferenceException(nameof(voltages));
+            }
+
+            if (times.Length != voltages.Length)
+            {
+                throw new InvalidOperationException("PWL - times array has different length than voltages array");
+            }
+
+            Points = CreatePoints(times, voltages, initialVoltage);
+            LineParameters = CreateLineParameters(Points);
```
Comment (svenboulanger) [inline SpiceSharp/Components/Waveforms/Pwl.cs]: I'd prefer not caching the line slopes and intercepts. If you really want to cache, you can do so when a breakpoint occurs. The parameters won't change until the next breakpoint anyway...

File: SpiceSharp/Components/Waveforms/Pwl.cs (hunk 2 above)
Comment (svenboulanger) [reply SpiceSharp/Components/Waveforms/Pwl.cs]: You can also calculate the slope and intercept here.

File: SpiceSharp/Components/Waveforms/Pwl.cs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        protected static double GetLineValue(List<Point> points, LineDefinition[] lines, double x)
+        {
+            int index = 0;
+
+            while (index < points.Count && points[index].X < x)
+            {
+                index++;
+            }
+
+            if (index == points.Count)
```
Comment (svenboulanger) [inline SpiceSharp/Components/Waveforms/Pwl.cs]: Put this inside the while loop and you can save a check.

File: SpiceSharp/Components/Waveforms/Pwl.cs [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+            {
+                return points[points.Count - 1].Y;
+            }
+
+            if (index == 0 && points[0].X > x)
```
Comment (svenboulanger) [inline SpiceSharp/Components/Waveforms/Pwl.cs]: Put this part (without `index == 0`) before the while loop and you can potentially save a check.

File: SpiceSharp/Components/Waveforms/Pwl.cs [hunk 7]
```diff
@@ -0,0 +1,155 @@
+﻿using System;
+using System.Collections.Generic;
+using SpiceSharp.IntegrationMethods;
+using SpiceSharp.Simulations;
+
+namespace SpiceSharp.Components.Waveforms
+{
+    public class Pwl : Waveform
+    {
+        public Pwl(double[] times, double[] voltages, double initialVoltage = 0.0)
```
[the remark this answers is not in the document]
Comment (marcin-golebiowski) [reply SpiceSharp/Components/Waveforms/Pwl.cs]: @svenboulanger  I found some docs of some SPICE simulator that were saying something different. But I think I will change it. I will make more sense.

File: SpiceSharp/Components/Waveforms/Pwl.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (marcin-golebiowski) [reply SpiceSharp/Components/Waveforms/Pwl.cs]: @svenboulanger Thanks to feedback. I'm checking with LTSpice now.

File: SpiceSharp/Components/Waveforms/Pwl.cs (hunk 2 above)
Comment (marcin-golebiowski) [reply SpiceSharp/Components/Waveforms/Pwl.cs]: @svenboulanger  For now I will go for creating all breakpoints only once. It has the advantage that waveform will have all the corners from pwl despite max step / step of transient.

Comment (svenboulanger) [reply SpiceSharp/Components/Waveforms/Pwl.cs]: The break point system already manages this. A breakpoint will force the time analysis to stop at a predetermined time point.

Comment (marcin-golebiowski) [reply SpiceSharp/Components/Waveforms/Pwl.cs]: @svenboulanger OK. I will try to change it. Please take a look at latests commits.

File: SpiceSharp/Components/Waveforms/Pwl.cs (hunk 4 above)
Comment (marcin-golebiowski) [reply SpiceSharp/Components/Waveforms/Pwl.cs]: @svenboulanger I will not cache.

File: SpiceSharp/Components/Waveforms/Pwl.cs [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+        public override void Setup()
+        {
+        }
+
+        protected List<Point> CreatePoints(double[] times, double[] voltages, double initialVoltage)
+        {
+            var result = new List<Point>();
+
+            for (var i = 0; i < times.Length; i++)
```
[the remark this answers is not in the document]
Comment (marcin-golebiowski) [reply SpiceSharp/Components/Waveforms/Pwl.cs]: @svenboulanger  I will add the check for monotonously increasing time points. There is something like that in LTSpice.

**Comment by marcin-golebiowski** [comment]
@svenboulanger  Is everything ok now?

**Pull request merged**